### PR TITLE
DistanceGeometry: add flag to enforce trans amides

### PR DESCRIPTION
**Pull request by greglandrum**
This adds the flag and makes it the default.

The code previously wasn't applying any constraints to amides, esters, etc.

Here's an example of a KDG conformer (no ET, which fixes this) of methyl acetamide from v2020.09.4.
image

With the fix we get:
image

and setting `forceTransAmides` to False gives:
image

### CODE REVIEW

File: Code/GraphMol/DistGeomHelpers/BoundsMatrixBuilder.h [hunk 1]
```diff
@@ -55,15 +55,17 @@ RDKIT_DISTGEOMHELPERS_EXPORT void initBoundsMat(DistGeom::BoundsMatPtr mmat,
 */
 RDKIT_DISTGEOMHELPERS_EXPORT void setTopolBounds(
     const ROMol &mol, DistGeom::BoundsMatPtr mmat, bool set15bounds = true,
-    bool scaleVDW = false, bool useMacrocycle14config = false);
+    bool scaleVDW = false, bool useMacrocycle14config = false,
```
Comment (bp-kelley) [inline Code/GraphMol/DistGeomHelpers/BoundsMatrixBuilder.h:58]: Getting close to needing a parameter object :)

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp [hunk 2]
```diff
@@ -880,10 +880,23 @@ bool setupInitialBoundsMatrix(
   unsigned int nAtoms = mol->getNumAtoms();
   if (params.useExpTorsionAnglePrefs || params.useBasicKnowledge) {
     setTopolBounds(*mol, mmat, etkdgDetails.bonds, etkdgDetails.angles, true,
-                   false, params.useMacrocycle14config);
+                   false, params.useMacrocycle14config,
+                   params.forceTransAmides);
   } else {
-    setTopolBounds(*mol, mmat, true, false, params.useMacrocycle14config);
+    setTopolBounds(*mol, mmat, true, false, params.useMacrocycle14config,
```
Comment (bp-kelley) [inline Code/GraphMol/DistGeomHelpers/Embedder.cpp:886]: There's the parameter object!

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp [hunk 3]
```diff
@@ -880,10 +880,23 @@ bool setupInitialBoundsMatrix(
...
+                   params.forceTransAmides);
   }
+#if 0
```
Comment (bp-kelley) [inline Code/GraphMol/DistGeomHelpers/Embedder.cpp:889]: Do you still want this?

File: Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp [hunk 4]
```diff
@@ -37,8 +37,8 @@ using namespace RDKit;
  * Version 2 taken from J. Chem. Inf. Model. 56, 1 (2016)
  *
  * torsion-angle potential form:
- * V = V1*(1 + s1*cos(1x)) + V2*(1 + s2*cos(2x)) + V3*(1 + s3*cos(1x))
- *     + V4*(1 + s4*cos(1x)) + V5*(1 + s5*cos(1x)) + V6*(1 + s6*cos(1x))
+ * V = V1*(1 + s1*cos(1x)) + V2*(1 + s2*cos(2x)) + V3*(1 + s3*cos(3x))
+ *     + V4*(1 + s4*cos(4x)) + V5*(1 + s5*cos(5x)) + V6*(1 + s6*cos(6x))
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp:41]: Or is this a doc change?

File: Code/GraphMol/DistGeomHelpers/BoundsMatrixBuilder.cpp [hunk 5]
```diff
@@ -815,12 +815,21 @@ bool _checkAmideEster14(const Bond *bnd1, const Bond *bnd3, const Atom *atm1,
...
+  //           << " bonds: " << bnd1->getIdx() << "," << bnd3->getIdx()
+  //           << std::endl;
+  // std::cerr << "   " << a1Num << " " << a3Num << " " << bnd3->getBondType()
+  //           << " " << a4Num << " " << bnd1->getBondType() << " " << a2Num <<
+  //           " "
+  //           << atm2->getTotalNumHs(true) << std::endl;
+  if (a3Num == 6 && bnd3->getBondType() == Bond::DOUBLE &&
       (a4Num == 8 || a4Num == 7) && bnd1->getBondType() == Bond::SINGLE &&
-      (a2Num == 8 || (a2Num == 7 && atm2->getTotalNumHs() == 1))) {
+      (a2Num == 8 || (a2Num == 7 && atm2->getTotalNumHs(true) == 1))) {
```
Comment (bp-kelley) [inline Code/GraphMol/DistGeomHelpers/BoundsMatrixBuilder.cpp:828]: Is this fix related to trans amids?

Comment (greglandrum) [reply Code/GraphMol/DistGeomHelpers/BoundsMatrixBuilder.cpp:828]: Yes, the fact that the wrong version of getTotalNumHs() was being called meant that some amides were not being correctly identified

File: Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp:41]: I don’t understand the question. This is a correction to a comment

[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp:41]: I meant should I find these coefficient changes in the code or whether it was a doc/comment change

[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp:41]: Just a doc change

File: Code/GraphMol/DistGeomHelpers/testDgeomHelpers.cpp [hunk 6]
```diff
@@ -2301,6 +2306,84 @@ void testGithub3667() {
...
+    params.forceTransAmides = true;
+    params.randomSeed = 0xf00d;
+    params.useExpTorsionAnglePrefs = false;
+    params.useBasicKnowledge = true;
+    auto cids = DGeomHelpers::EmbedMultipleConfs(*mol, 10, params);
+    for (auto cid : cids) {
+      TEST_ASSERT(cid >= 0);
+      auto conf = mol->getConformer(cid);
+      auto tors = MolTransforms::getDihedralDeg(conf, 0, 1, 3, 4);
+      TEST_ASSERT(fabs(fabs(tors) - 180) < 30);
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/DistGeomHelpers/testDgeomHelpers.cpp:2362]: Since this is int->double it's a safe conversion and there's no warning here.

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp (hunk 3 above)
Comment (greglandrum) [reply Code/GraphMol/DistGeomHelpers/Embedder.cpp:889]: nope, thanks for catching that

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/DistGeomHelpers/Embedder.cpp:886]: yeah, we could probably stand to refactor this code at some point.
The one defense I'll make is that `setTopolBounds()` isn't in the public API, so it's not like it'll confuse anyone else.

**Review by bp-kelley — APPROVED** [review]
